### DISCUSSION TITLE
Paragraph navigation skips over lines with only whitespace

### DISCUSSION THREAD

**Discussion by reinux** [JetBrains/ideavim#291] [General]
The other vim implementations I've seen -- vscode vim, vsvim, neovim, vim etc. -- all treat lines with only whitespace characters as being empty lines, so navigating using `{` and `}` will stop at these lines. Ideavim seems to jump over them.

Would be nice if this could be made to check for whitespace lines as well as blank lines, since a paragraph is visually a paragraph even if it has whitespace characters.

**Comment by reinux** [comment]
~~Hmm...~~

~~It looks like `MotionInnerParagraphAction` calls `getParagraphRange` with isOuter set to false (of course), which ultimately has the effect of rejecting blanks.~~

~~https://github.com/JetBrains/ideavim/blob/3b3fffe838c06c9339c32add1ceb1d38e77753f9/src/com/maddyhome/idea/vim/action/motion/object/MotionInnerParagraphAction.kt#L47~~

~~Is there a reason for this behavior that I'm missing?~~

Edit: Scratch that, just realized this is for the difference between `ip` and `ap`. Oddly, unlike `{` and `}`, `ip` and `ap` both treat lines with whitespaces as being blank.

**Comment by AlexPl292** [comment]
Hi! I'm a bit confused, as I see vim jumps over the line with empty spaces as well. Could you maybe attach some example of the text?
![2021-04-16 09 44 00](https://user-images.githubusercontent.com/4203721/114982673-6adcb980-9e98-11eb-95ca-8e617661e0f4.gif)

**Reply by reinux** [reply]
Sorry, I just realized this is the "correct" behavior that's consistent with vim (or at least nvim).

This is what I've gotten used to, and it seems more useful to me, since whitespaces are invisible and so jumping over them makes the behavior harder to predict.

From vsvim: 

https://user-images.githubusercontent.com/5127297/115039729-91452980-9e85-11eb-9758-8f5d7fd14bbc.mp4

It's also a little confusing, as using something like `vip` or `vap` in ideavim or nvim causes it to stop at the whitespace lines, even as it jumps over them when using `{ }`, which is seemingly inconsistent:

https://user-images.githubusercontent.com/5127297/115040022-d8331f00-9e85-11eb-9a8e-d4a6423e0006.mp4

**Comment by AlexPl292** [comment]
I've tried this in vsvim and it works now as in vim :) No idea if this was recently fixed.

Okay, in fact, I can't modify `}` behaviour because in vim docs it explicitly stands that

> Note that a blank line (only containing white space) is NOT a paragraph boundary.

But luckily, there are some plugins that can modify this behaviour and it was pretty easy to implement it for IdeaVim. So, I've added https://github.com/dbakker/vim-paragraph-motion plugin to IdeaVim and with the next release you can use modified `}` motion if you'll enable it using `set vim-paragraph-motion` command in your `~/.ideavimrc`: https://github.com/JetBrains/ideavim/wiki/Emulated-plugins#vim-paragraph-motion-to-be-released

Here is an issue in YouTrack: https://youtrack.jetbrains.com/issue/VIM-2290

**Reply by reinux** [reply]
Wow, thanks! I know this was kind of based on a false assumption I'd made, so I especially appreciate it.

Just tried vsvim, and it still stops on whitespace lines for me. I guess they may have updated it very recently, then.

**Reply by reinux** [reply]
Just tried this out. Much more comfortable for the way I use it.

Thanks a bunch!

**Reply by AlexPl292** [reply]
Awesome! Glad to hear it fits your workflow!
